---
title: "STUtility - A  Utility tool from the original Spatial Transcriptomics Research group"
date: ""
author: 
- Joseph BergenstrÃ¥hle, Royal Institute of Technology (KTH)
- Ludvig Larsson, Royal Institute of Technology (KTH)
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{STUtility vignette}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(STutility)

```


Introduction
============

STUtility is an Rpackage which includes various R-functions for interaction and analysis of Spatial Transcriptomics (ST) data. During the years of internal development of the ST method, the research group at SciLifeLab has used a vast array of R functions in their daily work. As the method is becoming widely adopted 


Pre-process and read in data
=============

STUtility provides two approached for reading in data and creating the S4 object. 

### Method 1

Info.table sheet to load in our samples togheter with meta data. 
The table looks like this:

```{r, echo=FALSE}

abs.path <- system.file("extdata", package = "STutility")

```

```{r}

infoTable <- read.table(paste(abs.path, "metaData.csv", sep="/"), sep=";", header=T)

head(infoTable)
```

The `samples` are the only mandatory column. However, if spotfiles and images are provided, the column headers of `spotfiles` and `imgs` also need to be named accordingly. 

With the info table as input, we create the S4 object which contains all the samples as well as meta data. 

```{r, echo=FALSE}
fix <-list.files(system.file("extdata/counts", package = "STutility"), full.names = T)
fixCounts <- fix[grep(fix, pattern = "[^_]MOB[1-9].tsv")]
fixSpotfiles <- fix[grep(fix, pattern = "alignment")]
fixImgs <- fix[grep(fix, pattern = ".jpg")]

infoTable[,1] <- fixCounts
infoTable[,5] <- fixSpotfiles
infoTable[,6] <- fixImgs

```

```{r}

cm <- prep.from.table(sampleTable=infoTable, 
                      transpose=T, 
                      topN=0, th.gene=0, th.spot=0, 
                      type="Seurat")

```

With the Seurat object created, we can proceed with the regular Seurat workflow to scale data, find variable features and conduct dimensionallity reductions:

```{r}

cm <- ScaleData(cm, vars.to.regress = c("nFeature_RNA", "sample"), verbose = T)
cm <- FindVariableFeatures(cm)
cm <- RunPCA(cm, ndims.print = 1, nfeatures.print = 5)
cm <- RunICA(cm, ndims.print = 1, nfeatures.print = 5)

```

For other alternatives, see the Seurat vignettes at https://satijalab.org/seurat/ . Every output is saved into the Seurat object. Visualization can be performed in the usual manner:

```{r, fig.width = 10, fig.height = 12}

cm <- SetIdent(cm, value = "sample")
DimPlot(object = cm, dims = 1:2)

```

Or, we can use the prefix "ST", if instead we would like to visualize the output onto the spatial array:

```{r, fig.width = 10, fig.height = 20}

ST.DimPlot(object = cm, 
           reduction = "pca",
           dims = 1:4, 
           group.by = "sample", 
           delim = "x|_", 
           palette = "RdBu",
           rev.cols = T,
           pt.size = 0.2,
           dark.theme = T, ncol = 4, grid.ncol = 1)

```



## QC 

```{r}
qcGenes(cm)



```





### Method 2



