---
title: "3D models on RA tissue"
author: "Ludvig Larsson"
date: "12/6/2019"
output:
  html_document:
    theme: darkly
    number_sections: false
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: true
---

<style type="text/css">
div.main-container {
  background-color: #000000 !important;
  max-width: 1400px;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style>
#TOC {
  background: url("https://www.spatialresearch.org/wp-content/uploads/2019/09/str-logo-spatial_research_3@2x.png");
  background-size: contain;
  padding-top: 100px !important;
  background-repeat: no-repeat;
  op: 5%;
  opacity: 0.8;
  width: 500px;
  color: white;
  border-color: #000000 !important;
}
</style>

<style> code, pre{
  background-color: #000000 !important;
  color: white !important;
}
</style>
<style> 
body {
  color: white
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load_libs, include=FALSE}
library(STutility)
library(zeallot)
se <- readRDS("~/10x/R_objects/se_RA")
```

```{r seurat, include=FALSE, eval=FALSE}

spotfiles <- list.files("~/RA_project/data/Patient2/HE_spots_xy_files", full.names = T)
xys <- lapply(spotfiles, function(f) {
  df <- read.table(file = f, header = T, sep = "\t", stringsAsFactors = F)
  xy <- data.frame(do.call(rbind, strsplit(rownames(df), split = "_")), stringsAsFactors = F)
  df <- setNames(cbind(xy, xy, df), nm = c("x", "y", "new_x", "new_y", "pixel_x", "pixel_y"))
  df <- as.data.frame(lapply(df, as.numeric))
})

dir.create(path = "~/RA_project/data/Patient2/spotfiles")
for (i in seq_along(xys)) {
  write.table(x = xys[i], file = paste0("~/RA_project/data/Patient2/spotfiles/selection_", i, ".tsv"), sep = "\t", row.names = F, col.names = T, quote = F)
}

```

```{r, include=FALSE, eval=FALSE}

ensids <- read.table("~/Fetal_heart_project/annotation/genes.tsv", header = T, stringsAsFactors = F)
ensids$gene_id <- substr(x = ensids$gene_id, start = 1, stop = 15)
ensids <- ensids[-which(duplicated(ensids$gene_id)), ]
write.table(x = ensids, file = "~/RA_project/genes.tsv", quote = F, sep = "\t", row.names = F, col.names = T)

```

```{r, include=FALSE, eval=FALSE}

samples <- list.files("~/RA_project/data/tsv", full.names = T)[5:11]
imgs <- list.files(pattern = ".jpg", "~/RA_project/data/Patient2/images", full.names = T)[1:7]
spotfiles <- list.files("~/RA_project/data/Patient2/spotfiles", full.names = T)

infoTable <- data.frame(samples, imgs, spotfiles, stringsAsFactors = FALSE)

```
<br>

# Read input

Here we have prepared a Seurat object containing data from RA tissue obtained from Carlberg et al (Scientific Reports, 2019), *"Exploring inflammatory signatures in arthritic joint biopsies with Spatial Transcriptomics"*. The data consists of 7 adjacent tissue sections collected from synovial biopsies from rheumatoid arthritis patients. The data was obtained using the 1000 spot arrays ('1k' platform). 
<br>
```{r seurat2, include=FALSE, eval=FALSE}

se <- InputFromTable(infoTable, 
                     platform = "1k", 
                     annotation = ensids, 
                     min.gene.count = 300, 
                     min.spot.count = 1e3)

```
<br>

# Load, mask and align images

We can load and mask the images using the `LoadImages` and `MaskImages` functions but the alignment has to be done manually using `ManualAlignImages`. 
<br>
```{r manual_align, eval=FALSE}

se <- LoadImages(se, time.resolve = F)
se <- MaskImages(se)
se <- ManualAlignImages(se, reference.index = 4)

```
<br>

## Check alignment

Now if we plot the tissue sections we can see that they are more or less aligned in the same orientation.
<br>
```{r plot_aligned, fig.width=12, fig.height=12}

ImagePlot(se, method = "raster", darken = TRUE)

```
<br>

# Normalize data

We'll use the `SCTransform` function from Seurat to normalize the data.
<br>
```{r normalization, eval=FALSE}

se@meta.data$sample_id <- paste0("sample_", se@tools$Staffli@meta.data$sample)

se <- SCTransform(se, vars.to.regress = c("sample_id"))

```
<br>

# Create 3D stack

To create the 3D model we run the `Create3DStack` function. `Create3DStack` will reload the images in higher resolution if necessary and try to segment out positions of nuclei in the tissue to generate aligned 2D point patterns. The 2D point patterns will then be stacked along a z axis.
<br>
```{r 3D_stack, eval = FALSE}

se <- Create3DStack(se)

```
<br>

## Find spatial features

We can run the `CorSpatialGenes` function to find genes with spatial autocorrelation.
<br>
```{r spatcor, warning=FALSE}

spatgenes <- CorSpatialGenes(se)
head(spatgenes, 20)

```
<br>

## Plotting in 3D

Now that we have a 3D model, we can interpolate any continuous feature values onto the 3D point cloud and visualize it. Here we downsample the number of points to 200,000 to make the rendering a bit smoother. 
<br>
```{r 3D_plot1}

FeaturePlot3D(se, features = "CCL21", pt.size = 1, cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), dark.theme = TRUE, pts.downsample = 2e5)

```

<br>
Sometimes it can be useful to add some opacity to the plot so the parts of the tissue with low feature values are more transparent. The `add.alpha` option can be activated to achieve this. We can also put the sections closer to each other along the z axis using the `add.margins` option.
<br>
```{r 3D_plot2}

FeaturePlot3D(se, features = "CCL21", pt.size = 0.5, cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), dark.theme = TRUE, pts.downsample = 2e5, add.alpha = TRUE, add.margins = 50)

```

<br>
And if you want to use the spots instead of the point cloud you can just set the `mode` option to "spots".
<br>
```{r 3D_plot3}

FeaturePlot3D(se, features = "CCL21", pt.size = 4, cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), dark.theme = TRUE,  mode = "spots")

```
<br>

# Non-negative Matrix Factorization
<br>
```{r nmf, eval=FALSE}

se <- RunNMF(se, nfactors = 10, n.cores = 7)

```
<br>

## Factor 1

CCL21 - Inhibits hemopoiesis and stimulates chemotaxis. Chemotactic in vitro for thymocytes and activated T-cells, but not for B-cells, macrophages, or neutrophils. 
CCL19 -  Recombinant CCL19 shows potent chemotactic activity for T-cells and B-cells but not for granulocytes and monocytes.
EVL - Ena/VASP proteins are actin-associated proteins involved in a range of processes dependent on cytoskeleton remodeling and cell polarity such as axon guidance and lamellipodial and filopodial dynamics in migrating cells.
CD3E + SELL CD4 T-cell markers.
<br>
```{r nmf1_plot, fig.width=10, fig.height=10, out.width="100%"}

cowplot::plot_grid(
  ST.DimPlot(se, reduction = "NMF", dims = 1, 
             pt.size = 1, dark.theme = T, ncol = 2,
             cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), 
             center.zero = F),
  FactorGeneLoadingPlot(se, factor = 1, dark.theme = T) + theme(plot.background = element_rect(fill = "black", colour = "black")),
  ncol = 2)

```
<br>

## Factor 2

I think this factor must be some artefact along the edges. Maybe the cells have initiated a stress response or something ...
<br>
```{r nmf2_plot, fig.width=10, fig.height=10, out.width="100%"}

cowplot::plot_grid(
  ST.DimPlot(se, reduction = "NMF", dims = 2, 
             pt.size = 1, dark.theme = T, ncol = 2,
             cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), 
             center.zero = F),
  FactorGeneLoadingPlot(se, factor = 2, dark.theme = T) + theme(plot.background = element_rect(fill = "black", colour = "black")),
  ncol = 2)

```
<br>
## Factor 3

MZB1 = B/plasma cell marker 
<br>
```{r nmf3_plot, fig.width=10, fig.height=10, out.width="100%"}

cowplot::plot_grid(
  ST.DimPlot(se, reduction = "NMF", dims = 3, 
             pt.size = 1, dark.theme = T, ncol = 2,
             cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), 
             center.zero = F),
  FactorGeneLoadingPlot(se, factor = 3, dark.theme = T) + theme(plot.background = element_rect(fill = "black", colour = "black")),
  ncol = 2)

```
<br>
## Factor 4

MZB1 B cell marker, HBA1/HBA2 blood markers
<br>
```{r nmf4_plot, fig.width=10, fig.height=10, out.width="100%"}

cowplot::plot_grid(
  ST.DimPlot(se, reduction = "NMF", dims = 4, 
             pt.size = 1, dark.theme = T, ncol = 2,
             cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), 
             center.zero = F),
  FactorGeneLoadingPlot(se, factor = 4, dark.theme = T) + theme(plot.background = element_rect(fill = "black", colour = "black")),
  ncol = 2)

```
<br>

## Factor 5

CD79A/CD79B - B lymphocyte antigen receptor
CD52 - Present on mature lymphocytes
<br>
```{r nmf5_plot, fig.width=10, fig.height=10, out.width="100%"}

cowplot::plot_grid(
  ST.DimPlot(se, reduction = "NMF", dims = 5, 
             pt.size = 1, dark.theme = T, ncol = 2,
             cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), 
             center.zero = F),
  FactorGeneLoadingPlot(se, factor = 5, dark.theme = T) + theme(plot.background = element_rect(fill = "black", colour = "black")),
  ncol = 2)

```
<br>

## Factor 6

Ribosomal protein genes ...
<br>
```{r nmf6_plot, fig.width=10, fig.height=10, out.width="100%"}

cowplot::plot_grid(
  ST.DimPlot(se, reduction = "NMF", dims = 6, 
             pt.size = 1, dark.theme = T, ncol = 2,
             cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), 
             center.zero = F),
  FactorGeneLoadingPlot(se, factor = 6, dark.theme = T) + theme(plot.background = element_rect(fill = "black", colour = "black")),
  ncol = 2)

```
<br> 

## Factor 7

C1QB/C1QC - Antigen presenting cells, e.g. Denritic cells macrophages?
CFD - This protease catalyzes the cleavage of factor B, the rate-limiting step of the alternative pathway of complement activation.
CD14 - (cluster of differentiation 14) is a human protein made mostly by macrophages as part of the innate immune system
<br>
```{r nmf7_plot, fig.width=10, fig.height=10, out.width="100%"}

cowplot::plot_grid(
  ST.DimPlot(se, reduction = "NMF", dims = 7, 
             pt.size = 1, dark.theme = T, ncol = 2,
             cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), 
             center.zero = F),
  FactorGeneLoadingPlot(se, factor = 7, dark.theme = T) + theme(plot.background = element_rect(fill = "black", colour = "black")),
  ncol = 2)

```
<br>

## Factor 8

CD6 on surface of T-cells
SIGIRR - Acts as a negative regulator of the Toll-like and IL-1R receptor signaling pathways
<br>
```{r nmf8_plot, fig.width=10, fig.height=10, out.width="100%"}

cowplot::plot_grid(
  ST.DimPlot(se, reduction = "NMF", dims = 8, 
             pt.size = 1, dark.theme = T, ncol = 2,
             cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), 
             center.zero = F),
  FactorGeneLoadingPlot(se, factor = 8, dark.theme = T) + theme(plot.background = element_rect(fill = "black", colour = "black")),
  ncol = 2)

```
<br>

## Factor 9

HLA-DRA/HLA-DBQ1 - Class II molecules are expressed in antigen presenting cells (APC: B lymphocytes, dendritic cells, macrophages)
TMSB4X - This gene encodes an actin sequestering protein which plays a role in regulation of actin polymerization. The protein is also involved in cell proliferation, migration, and differentiation.
VIM - This gene encodes a type III intermediate filament protein. Intermediate filaments, along with microtubules and actin microfilaments, make up the cytoskeleton. Vimentins are class-III intermediate filaments found in various non-epithelial cells, especially mesenchymal cells.
MS4A1 - B cell marker
<br>
```{r nmf9_plot, fig.width=10, fig.height=10, out.width="100%"}

cowplot::plot_grid(
  ST.DimPlot(se, reduction = "NMF", dims = 9, 
             pt.size = 1, dark.theme = T, ncol = 2,
             cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), 
             center.zero = F),
  FactorGeneLoadingPlot(se, factor = 9, dark.theme = T) + theme(plot.background = element_rect(fill = "black", colour = "black")),
  ncol = 2)

```
<br>

## Factor 10

HLA-DRB1 - Class II molecules are expressed in antigen presenting cells (APC: B lymphocytes, dendritic cells, macrophages).
CD14 - (cluster of differentiation 14) is a human protein made mostly by macrophages as part of the innate immune system

<br>
```{r nmf10_plot, fig.width=10, fig.height=10, out.width="100%"}

cowplot::plot_grid(
  ST.DimPlot(se, reduction = "NMF", dims = 10, 
             pt.size = 1, dark.theme = T, ncol = 2,
             cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"), 
             center.zero = F),
  FactorGeneLoadingPlot(se, factor = 10, dark.theme = T) + theme(plot.background = element_rect(fill = "black", colour = "black")),
  ncol = 2)

```
<br>

## HSVPlot
<br>
```{r hsv2d, fig.width=18, fig.height=10, out.width="100%", warning=FALSE}

p1 <- HSVPlot(se, features = paste0("factor_", c(1, 10, 5, 7)), pt.size = 1.5)
c(df1, df2) %<-% SummarizeAssocFeatures(se, dims = c(1, 10, 5, 7))

p2 <- swne::ggHeat(df2[nrow(df2):1, ], rescaling = "column") +
  scale_fill_gradientn(colours = viridis::magma(n = 11)) +
  DarkTheme() +
  theme(plot.background = element_rect(fill = "black", colour = "black"), plot.margin = margin(2, 2, 2, 2, "cm"))

cowplot::plot_grid(p1, p2, ncol = 2)

```
<br>

## Plot 3D

We can also plot dimensionality reduction output in 3D using the `DimPlot3D` function.
<br>
```{r NMF_3D}

DimPlot3D(se, reduction = "NMF", dims = 7, dark.theme = T, cols = c("black", "navyblue", "cyan", "yellow", "red", "dark red"))

```
<br>

## HSVPlot 3D

Or multiple dimensionality reduction vectors in 3D using the `HSVPlot3D` function.
<br>
```{r HSV_NMF_3D}

HSVPlot3D(se, features = paste0("factor_", c(1, 10, 5, 7)), dark.theme = T)

```
<br>

# Date
<br>
```{r date}
date()
```
<br>

# Session Info
<br>
```{r sessioninfo}
sessionInfo()
```



